### ISSUE TITLE
Authentication - add Google OAuth2

### ISSUE THREAD

**Issue by waybeforenow** [latenighttales/alcali#119]
Similarly to #83, it would be really great if we could get Google Account integration in Alcali via OAuth2. We could use an account- or domain-based whitelist to authenticate users.

If the Google OAuth2 module is enabled, the entire `Login` view can be replaced with a redirect to the authentication URL, then the authorization code from the callback can be passed from the backend and used to request an access token. Once the access token is retrieved, it can be discarded since we only care about verifying the identity of the user (which is contained in the `id_token`).

Pointers:
* [Overview of OAuth2 authentication flow](https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2)
* [Google OAuth2 example using `requests-oauthlib`](https://requests-oauthlib.readthedocs.io/en/latest/examples/google.html)

There is the tricky question of how to determine which users have admin rights. For my purposes, granting admin privileges to any whitelisted user is fine. I'd be interested to hear your thoughts on this.

**Comment by mattLLVW** [comment]
Thanks for the issue!

It should be doable, and any OAuth providers from [django-allauth](https://www.intenct.nl/projects/django-allauth/) could be used.
It seems that we could load auth providers dynamically so that's nice.

> There is the tricky question of how to determine which users have admin rights. 

Right now, admin rights are barely used(create users and display general settings iirc). If AllAuth create the user, Admin rights could be granted from the Users view. 

> We could use an account- or domain-based whitelist to authenticate users.

Looks simple enough.(per [this](https://stackoverflow.com/questions/19113623/django-allauth-only-allow-users-from-a-specific-google-apps-domain) post).

I'll fiddle with it a bit and see what can be done.

**Comment by waybeforenow** [comment]
Hey, just checking in. I have an update on my use-case.

I started putting [`nginx-google-oauth`](https://github.com/cloudflare/nginx-google-oauth) in front of my Alcali install. This means that I can now take advantage of Django's [`RemoteUserMiddleware`](https://docs.djangoproject.com/en/2.2/howto/auth-remote-user/) to bypass the login screen entirely and authenticate users based on their GSuite (Google login) principal.

For instance, if a user `foo@example.com` logs into the nginx reverse proxy, this additional header will be passed to the backend:

```
REMOTE_USER: foo
```

Based on the documentation, I think this might be much easier to implement than `django-allauth`. The important part is making sure users don't turn it on unless they have a reverse proxy in place that strips the `REMOTE_USER` header from a request before it hits the backend. Otherwise logins can be easily spoofed.

I'm happy to test any code you want to throw at me; let me know.

**Comment by mattLLVW** [comment]
Hey, so yeah, django-allauth is not the way to go.
i got something working with [vue-google-oauth2](https://github.com/guruahn/vue-google-oauth2) for the frontend and [django-rest-social-auth](https://github.com/st4lk/django-rest-social-auth).

> I'm happy to test any code you want to throw at me; let me know.

I'll push a branch soon(most likely this weekend), if you can give some feedback it would be great !

**Comment by mattLLVW** [comment]
Hey @waybeforenow , i pushed #130 .

Here are the environment variable you should set:
```bash
AUTH_BACKEND=social
SOCIAL_AUTH_GOOGLE_OAUTH2_KEY=XXXXX.apps.googleusercontent.com
SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET=XXX
SOCIAL_AUTH_REDIRECT_URI=<FULL URI> ex: https://foo.bar:9000
```
To limit access to certain emails:
```bash
SOCIAL_AUTH_GOOGLE_OAUTH2_WHITELISTED_EMAILS=<COMMA SEP EMAILS>
```
and/or certain domains:
```bash
SOCIAL_AUTH_GOOGLE_OAUTH2_WHITELISTED_DOMAINS=<COMMA SEP DOMAINS>
```

Don't forget to [apply migrations](https://alcali.dev/running/#apply-migrations).

**Comment by mattLLVW** [comment]
Merged and i created a release (2019.2.3)
